analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs

```diff
@@ -168,133 +144,275 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
+    int? MultiplicationAndAssignments2(int? i) => (i = null) * (i = 42) * i.Value; // Compliant
+}
 
-public virtual DateTime? Foo
+class ComplexConditionsSingleNullable
 {
-    get => foo;
-    set
-    {
-        if (value.HasValue)
-        {
-            //HasValue and NoValue constraints are set here
-        }
-        if (foo != value || foo.HasValue)
-        {
-            foo = value;
-        }
-    }
+    bool LogicalAndLearningNonNull1(bool? b) => b.HasValue && b.Value;
+    bool LogicalAndLearningNonNull2(bool? b) => b.HasValue == true && b.Value;
+    bool LogicalAndLearningNonNull3(bool? b) => b.HasValue != false && b.Value;
+    bool LogicalAndLearningNonNull4(bool? b) => b.HasValue == !false && b.Value;
+    bool LogicalAndLearningNonNull5(bool? b) => !!b.HasValue && b.Value;
+
+    bool LogicalAndLeaningNull1(bool? b) => !b.HasValue && b.Value;         // Noncompliant - parameter should be null-constrained
+    bool LogicalAndLeaningNull2(bool? b) => b.HasValue == false && b.Value; // Noncompliant
+    bool LogicalAndLeaningNull3(bool? b) => b.HasValue == !true && b.Value; // Noncompliant
+    bool LogicalAndLeaningNull4(bool? b) => !!!b.HasValue && b.Value;       // Noncompliant
+
+    bool BitAndLearningNull(bool? b) => !b.HasValue & b.Value;              // Noncompliant - i.Value reached in any case
+    bool BitAndLearningNonNull(bool? b) => b.HasValue & b.Value;            // Noncompliant - i.Value reached in any case
+
+    bool LogicalOrLearningNonNull(bool? b) => b.HasValue || b.Value;        // Noncompliant - i.Value reached when i.HasValue is false
+
+    bool BitOrLearningNonNull(bool? b) => b.HasValue | b.Value;             // Noncompliant - i.Value reached in any case
+
+    bool Tautology(bool? b) => (!b.HasValue || b.HasValue) && b.Value;      // Noncompliant
+
+    bool ShortCircuitedOr(bool? b) => (true || b.HasValue) && b.Value;      // Compliant
+    bool ShortCircuitedAnd(bool? b) => (false &&  b.HasValue) || b.Value;   // Compliant
+
+    bool XorWithTrue(bool? b) => (true ^ b.HasValue) && b.Value;            // Noncompliant
+    bool XorWithFalse(bool? b) => (false ^ b.HasValue) && b.Value;          // Compliant
+}
+
+class ComplexConditionMultipleNullables
+{
+    bool IndependentConditions1(double? d, float? f) => f == null && d.Value == 42.0;  // Compliant, f imposes no constraints on d
+    bool IndependentConditions2(double? d, float? f) => f != null && d.Value == 42.0;  // Compliant
+    bool IndependentConditions3(double? d, float? f) => f.HasValue && d.Value == 42.0; // Compliant
+    bool IndependentConditions4(double? d, float? f) => null == f && d.Value == 42.0;  // Compliant
+    bool IndependentConditions5(double? d, float? f) => null == f && d.Value == 42.0;  // Compliant
+
+    bool DependentConditions1(double? d, float? f) =>
+        !d.HasValue && d.Value == 42.0;                                 // Noncompliant
+    bool DependentConditions2(double? d, float? f) =>
```

Comment:
Unused
```suggestion
    bool DependentConditions1(double? d) =>
        !d.HasValue && d.Value == 42.0;                                 // Noncompliant
    bool DependentConditions2(double? d) =>
```

Reply:
The idea here was to test that the mere presence of a second nullable value doesn't affect the analysis of the first nullable value. The class is called `ComplexConditionMultipleNullables` and methods are called `DependentConditions`, where the most obvious way in which two variable can depend on each other is them being independent.

To make all this more explicit, I have:
- moved `DependentConditions2` to `ComplexConditionsSingleNullable` and changed into `bool LogicalAndLearningNonNull6(bool? b) => b.Value && d.HasValue;`. As you pointed out, having `b.Value` first is just another way of learning that `b` has a value.
- made more explicit `DependentConditions1` comment: `// Noncompliant, f presence doesn't affect d`.
- renumbered following dependent condition cases.